src/macports1.0/macports.tcl

```diff
@@ -2102,6 +2102,7 @@ proc _mportexec {target mport} {
     set workername [ditem_key $mport workername]
     $workername eval {validate_macportsuser}
     if {![catch {$workername eval "check_variants $target"} result] && $result == 0 &&
+        ![catch {$workername eval {_check_xcode_version}} result] && $result == 0 &&
```

Comment:
If you're going to use _check_xcode_version in this way, it needs to check _archive_available first.

Reply:
I've moved the _check_xcode_version call to portfetch now instead of _mportexec, that should work right?

Reply:
No, Xcode is not required to fetch or extract. I'd recommend going back to the previous version of the change, but adding this at the start of _check_xcode_version:
```
if {[_archive_available]} {
    return 0
}
```

Reply:
Right, I've done that. When is _archive_available used before?

Reply:
> When is _archive_available used before?

Several places, for example in the calculation of dependencies, since if there's an archive then build-time deps are not needed.

Further on the point that Xcode is not needed to fetch, I think that _target_needs_deps is no longer a sufficient check. We really need another check for whether the target needs a toolchain.

Reply:
@jmroot Would you suggest adding

```tcl
proc macports::_target_needs_toolchain {target} {
    switch -- $target {
         patch -
         configure -
         build -
         test -
         destroot -
         install -
         activate -
         dmg -
         mdmg -
         pkg -
         mpkg {
              return 1
         }
         default {
              return 0
         }
    }
}
```

instead then and replacing `_target_needs_deps $target` in the beginning of `mportexec`, plus guarding the `_check_xcode_version` check in `_mportexec` with a `if {_target_needs_toolchain $target}` then?

In any case, the current behavior is definitely wrong and needs to be fixed, and this patch is better in requiring Xcode a little too often rather than not requiring it often enough, so I'm going to merge it as-is now. Please let me know whether my proposal outlined above is what you had in mind – I'll follow-up with a patch then.

Reply:
In _mportexec, the target is always activate, so the Xcode check is always needed if an archive is not available. In mportexec, the target may need Xcode even if an archive is available (e.g. build). So there need to be three cases in _target_needs_toolchain: needs toolchain, doesn't need toolchain, and needs toolchain if there is no archive.

So the _archive_available check needs to be moved outside of _check_xcode_version, and the new proc would look more like this:
```tcl
proc macports::_target_needs_toolchain {target} {
    switch -- $target {
         configure -
         build -
         test -
         destroot {
             return yes
         }
         install -
         activate -
         dmg -
         mdmg -
         pkg -
         mpkg {
              return if_no_archive
         }
         default {
              return no
         }
    }
}
```
I excluded patch as well because /usr/bin/patch comes with the base OS. Technically a portfile could do something that needs a toolchain in post-extract etc., but I think it's better to err on the side of allowing those targets to try to run.

Reply:
Right, that sounds good. I'll try to follow-up with this change soonish.

Reply:
See #143.